File: 2011m./uzd2./sprendimai/uzd2.cpp
#include <iostream>
#include <vector>
#include <fstream>
#include <string>
#include <algorithm>
#include <numeric>
#include <sstream>

using namespace std;

struct Performance {
    string name;
    vector<int> technique;
    vector<int> artistry;
    int score;

    Performance(string n, vector<int> t, vector<int> a) : name(n), technique(t), artistry(a), score(0) {}
};

vector<Performance> sortPerformances(vector<Performance>& performances) {
    sort(performances.begin(), performances.end(), [](const Performance& a, const Performance& b) {
        return a.score > b.score;
    });
    return performances;
}

int countScore(vector<int>& scores) {
    auto minMax = minmax_element(scores.begin(), scores.end());
    scores.erase(minMax.first);
    scores.erase(find(scores.begin(), scores.end(), *minMax.second));
    return accumulate(scores.begin(), scores.end(), 0);
}

int main() {
    ifstream input("U2.txt");
    ofstream output("U2rez.txt");

    int contestantCount;
    string judgeCount;
    input >> contestantCount >> judgeCount;
    
    vector<string> contestants;
    vector<Performance> performances;
    
    input.ignore();

    while (contestants.size() < contestantCount) {
        string teamName;
        getline(input, teamName);

        istringstream issNames(teamName);
        string name1, name2;
        issNames >> name1 >> name2;

        if (!name1.empty() && find(contestants.begin(), contestants.end(), name1) == contestants.end()) {
            contestants.push_back(name1);
        }
        if (!name2.empty() && find(contestants.begin(), contestants.end(), name2) == contestants.end()) {
            contestants.push_back(name2);
        }

        string line;
        vector<int> technique, artistry;

        getline(input, line);
        istringstream issTechnique(line);
        int score;
        while (issTechnique >> score) {
            technique.push_back(score);
        }

        getline(input, line);
        istringstream issArtistry(line);
        while (issArtistry >> score) {
            artistry.push_back(score);
        }

        performances.emplace_back(teamName, technique, artistry);
    }

    for (auto& performance : performances) {
        int score = countScore(performance.technique);
        score += countScore(performance.artistry);
        performance.score = score;
    }

    performances = sortPerformances(performances);

    for (const auto& performance : performances) {
        output << performance.name;
        output << string(20 - performance.name.size(), ' ');
        output << performance.score << endl;
    }

    input.close();
    output.close();

    return 0;
}

File: 2011m./uzd2./sprendimai/uzd2.py
class Performance:
    score = 0
    def __init__(self, name, technique, artistry) -> None:
        self.name = name
        self.technique = technique
        self.artistry = artistry

def sortPerformances(performances):
    for i in range(len(performances)):
        for j in range(i + 1, len(performances)):
            if(performances[i].score < performances[j].score):
                performances[i], performances[j] = performances[j], performances[i]
    return performances

def countScore(scores):
    scores.remove(min(scores))
    scores.remove(max(scores))
    return sum(scores)

def main():
    input = open("U2.txt", "r")

    data = next(input).split(" ")

    contestantCount = int(data[0])
    judgeCount = data[1]
    contestants = []
    performances = []
    
    while(len(contestants) < contestantCount):
        teamName = next(input)
        teamName = teamName[:-2]
        names = teamName.split(" ")

        if(names[0] not in contestants):
            contestants.append(data[0])
        if(names[1] not in contestants):
            contestants.append(data[1])
        
        technique = next(input).split(" ")
        technique = [int(o) for o in technique]
        artistry = next(input).split(" ")
        artistry = [int(o) for o in artistry]

        performances.append(Performance(teamName, technique, artistry))
    
    for performance in performances:
        score = countScore(performance.technique)
        score += countScore(performance.artistry)
        performance.score = score
    
    performances = sortPerformances(performances)

    output = open("U2rez.txt", "w")

    for performance in performances:
        output.write(f'{performance.name}{" "*(20-len(performance.name))}{performance.score}\n')

    input.close()
    output.close()

if __name__ == "__main__":
    main()